Modules/_datetimemodule.c

```diff
@@ -3259,21 +3259,37 @@ date_fromtimestamp(PyObject *cls, PyObject *obj)
 static PyObject *
 date_today(PyObject *cls, PyObject *Py_UNUSED(dummy))
 {
-    PyObject *time;
-    PyObject *result;
-    time = time_time();
-    if (time == NULL)
-        return NULL;
+    /* Use C implementation to boost performance for date type */
+    if ((PyTypeObject *)cls == &PyDateTime_DateType) {
+        struct tm tm;
+        time_t t;
+        time(&t);
 
-    /* Note well:  today() is a class method, so this may not call
-     * date.fromtimestamp.  For example, it may call
-     * datetime.fromtimestamp.  That's why we need all the accuracy
-     * time.time() delivers; if someone were gonzo about optimization,
-     * date.today() could get away with plain C time().
-     */
-    result = PyObject_CallMethodOneArg(cls, &_Py_ID(fromtimestamp), time);
-    Py_DECREF(time);
-    return result;
+        if (_PyTime_localtime(t, &tm) != 0) {
+            return NULL;
+        }
+
+        return new_date_ex(tm.tm_year + 1900,
+                           tm.tm_mon + 1,
+                           tm.tm_mday,
+                           (PyTypeObject*)cls);
+    }
+    else {
+        PyObject *time;
+        PyObject *result;
+        time = time_time();
+        if (time == NULL) {
+            return NULL;
+        }
+
+        /* Note well:  today() is a class method, so this may not call
+         * date.fromtimestamp.  For example, it may call
+         * datetime.fromtimestamp.
+         */
+        result = PyObject_CallMethodOneArg(cls, &_Py_ID(fromtimestamp), time);
+        Py_DECREF(time);
+        return result;
```

Comment:
```suggestion
    else {
        PyObject *time = time_time();
        if (time == NULL) {
            return NULL;
        }
        /* Note well: since today() is a class method, it may not call
         * date.fromtimestamp, e.g., it may call datetime.fromtimestamp.
         */
        PyObject *res = PyObject_CallMethodOneArg(cls, &_Py_ID(fromtimestamp), time);
        Py_DECREF(time);
        return res;
```

We can reduce a bit the number of lines. Or, to even further reduce the diff, you can just have:

```c
if (...) {
   // fast path
   return ...
}

// keep previous code but update comment
```